packages/playwright-test/src/reporters/blob.ts

```diff
@@ -0,0 +1,184 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { EventEmitter } from 'events';
+import fs from 'fs';
+import path from 'path';
+import { ManualPromise, ZipFile } from 'playwright-core/lib/utils';
+import { yazl } from 'playwright-core/lib/zipBundle';
+import type { FullConfig, FullResult } from '../../types/testReporter';
+import type { Suite } from '../common/test';
+import type { FullConfigInternal } from '../common/types';
+import { type JsonEvent, type JsonProject, type JsonSuite, TeleReporterReceiver } from '../isomorphic/teleReceiver';
+import HtmlReporter, { defaultReportFolder } from './html';
+import { TeleReporterEmitter } from './teleEmitter';
+
+
+type BlobReporterOptions = {
+  outputDir?: string;
+};
+
+export class BlobReporter extends TeleReporterEmitter {
+  private _messages: any[] = [];
+  private _options: BlobReporterOptions;
+  private _outputFile: string | undefined;
+
+  constructor(options: BlobReporterOptions = {}) {
+    super(message => this._messages.push(message));
+    this._options = options;
+  }
+  override onBegin(config: FullConfig<{}, {}>, suite: Suite): void {
+    super.onBegin(config, suite);
+    const outputDir = this._resolveOutputDir(config as FullConfigInternal);
+    let shardSuffix = '';
+    if (config.shard) {
+      const paddedNumber = String(config.shard.current).padStart(String(config.shard.total).length, '0');
+      shardSuffix = `-${paddedNumber}-of-${config.shard.total}`;
+    }
+    this._outputFile = path.join(outputDir, `report${shardSuffix}.zip`);
+  }
+
+  override async onEnd(result: FullResult): Promise<void> {
+    await super.onEnd(result);
+    if (!this._outputFile)
```

Comment:
How can this happen? Looks like we should have thrown when calculating the output file?